diff --git a/WORKSPACE b/WORKSPACE
index 19c30f7e..8a2e0770 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -32,6 +32,7 @@ http_archive(
     name = "boringssl",
     patch_cmds = [
         """sed -i.bak 's/bio.c",/bio.c","src\\/decrepit\\/bio\\/base64_bio.c",/g' BUILD.generated.bzl""",
+	"""sed -i.bak 's/-Werror//g' BUILD"""
     ],
     sha256 = "a9c3b03657d507975a32732f04563132b4553c20747cec6dc04de475c8bdf29f",
     strip_prefix = "boringssl-80ca9f9f6ece29ab132cce4cf807a9465a18cfac",
